go-controller/pkg/node/gateway_iptables.go

```diff
@@ -326,7 +326,12 @@ func recreateIPTRules(table, chain string, keepIPTRules []iptRule) {
 	}
 }
 
-func getGatewayIPTRules(service *kapi.Service, nodePortTargetIP string, nodeIP *net.IPNet) []iptRule {
+// getGatewayIPTRules returns NodePort and ExternalIP iptables rules for service. If nodeIP is non-nil, then
+// only incoming traffic on that IP will be accepted for NodePort rules; otherwise incoming traffic on the NodePort
+// on all IPs will be accepted. If internalLoadBalancerIP is "", then NodePort traffic will be redirected to
+// the service port on the service's ClusterIP. Otherwise, it will be redirected to the NodePort on the
+// internalLoadBalancerIP.
+func getGatewayIPTRules(service *kapi.Service, internalLoadBalancerIP string, nodeIP *net.IPNet) []iptRule {
```

Comment:
I wouldn't call this internalLoadBalancerIP; that's a little ambiguous. It's really an ovn gateway router IP.

Reply:
@girishmg @danwinship why does shared gateway mode behave like this? Is it because of the br-local thing in shared gw mode? For local gw (assuming a node ip of 1.1.1.1, GR of 169.254.0.2) we have packet destined for  1.1.1.1 node port. DNAT in host from 1.1.1.1-> 169.254.0.2. Packet arrives at GR, then DNAT's directly to endpoint (never goes through cluster IP).

Can someone describe the shared behavior?

Reply:
Here is a bit of context from a discussion I had with @girishmg : https://ovn-org.slack.com/archives/C010SQ5FSNL/p1591295321264600

I don't think it answers all questions, but might help fill in some blanks

Reply:
Thanks. After talking with @girishmg in shared gw mode we DNAT from host directly to cluster IP, then route cluster IP to the distributed router and SNAT via MP0 (to guarantee return traffic path). This is because the GR has the same IP as the node, so this makes sense...for now. In the future we need to look at all of these paths into the data plane and see if we can consolidate them.